domain/src/main/java/org/oppia/android/domain/topic/TopicListController.kt

```diff
@@ -161,101 +190,291 @@ class TopicListController @Inject constructor(
+      val storyId = storyProgress.storyId
+      val story = topicController.retrieveStory(topic.topicId, storyId)
+      val isStoryCompleted = checkIfStoryIsCompleted(storyProgress, story)
+      if (isStoryCompleted) {
+        return true
+      }
+    }
+    return false
+  }
+
+  private fun checkIfStoryIsCompleted(
+    storyProgress: StoryProgress,
+    storySummary: StorySummary
+  ): Boolean {
+    val completedChapterProgressList = getCompletedChapterProgressList(storyProgress)
+    val lastChapterSummary = storySummary.chapterList.lastOrNull()
+    return completedChapterProgressList.find { chapterProgress ->
+      chapterProgress.explorationId == lastChapterSummary?.explorationId
+    } != null
+  }
+
+  private fun getStartedChapterProgressList(storyProgress: StoryProgress): List<ChapterProgress> =
+    getSortedChapterProgressListByPlayState(
+      storyProgress, playState = ChapterPlayState.STARTED_NOT_COMPLETED
+    )
+
+  private fun getCompletedChapterProgressList(storyProgress: StoryProgress): List<ChapterProgress> =
+    getSortedChapterProgressListByPlayState(
+      storyProgress, playState = ChapterPlayState.COMPLETED
+    )
+
+  private fun getSortedChapterProgressListByPlayState(
+    storyProgress: StoryProgress,
+    playState: ChapterPlayState
+  ): List<ChapterProgress> {
+    return storyProgress.chapterProgressMap.values
+      .filter { chapterProgress -> chapterProgress.chapterPlayState == playState }
+      .sortedByDescending { chapterProgress -> chapterProgress.lastPlayedTimestamp }
+  }
+
+  private fun createOngoingStoryListBasedOnRecentlyPlayed(
+    storyId: String,
+    story: StorySummary,
+    recentlyPlayerChapterProgress: ChapterProgress,
+    completedChapterProgressList: List<ChapterProgress>,
+    topic: Topic,
+    isTopicConsideredCompleted: Boolean
+  ): Pair<PromotedStory?, Long> {
+    var numberOfDaysPassed = 0L
+    val recentlyPlayerChapterSummary: ChapterSummary? =
+      story.chapterList.find { chapterSummary ->
+        recentlyPlayerChapterProgress.explorationId == chapterSummary.explorationId
+      }
+    if (recentlyPlayerChapterSummary != null) {
+      numberOfDaysPassed = recentlyPlayerChapterProgress.getNumberOfDaysPassed()
+      return Pair(
+        addPromotedStoryInRecommendedStoryList(
+          storyId,
+          topic,
+          completedChapterProgressList.size,
+          story.chapterCount,
+          recentlyPlayerChapterSummary.name,
+          recentlyPlayerChapterSummary.explorationId,
+          isTopicConsideredCompleted
+        ),
+        numberOfDaysPassed
+      )
+    }
+    return Pair(null, numberOfDaysPassed)
+  }
+
+  private fun createOngoingStoryListBasedOnMostRecentlyCompleted(
+    storyId: String,
+    story: StorySummary,
+    mostRecentCompletedChapterProgress: ChapterProgress,
+    completedChapterProgressList: List<ChapterProgress>,
+    topic: Topic,
+    isTopicConsideredCompleted: Boolean
+  ): Pair<PromotedStory?, Long> {
```

Comment:
For this method & above: do we even need to create the promoted story object? It's not used by upstream code. From my perspective, it seems like this & the above method can just be retrofitted to return a ``Long`` indicating the number of days since the story was last played. That will simplify these two methods and upstream code.

Reply:
We need PromotedStory object that is getting added up into the PromotedStoryList.

Reply:
I misread the ``let`` above. I actually think there are a couple ways this can be further simplified:
1. You can move the PromotedStory creation to the callsite. Since it's completely duplicated, we should try to create it in just one location at the callsite if possible.
2. You could pass the filter down into this function & check it against the computed days count & return a boolean whether the filter matches & the other conditions are met (where currently it would return null, in the new function it would return false).

The above consolidates the code a bit more & avoids the need of using Pair at all which should be a simplification. The methods should also be renamed to better capture what they're doing with this change.

Reply:
Done.